skglm/experimental/pdcd_ws.py

```diff
@@ -0,0 +1,242 @@
+    when performing a primal update::
+
+        \text{score}_j = \abs{w_j - prox_{\tau_j, G_j}(w_j - \tau_j <X_j, z>)}
+
+    where :maths:`\tau_j` is the primal step associated with the j-th feature.
+
+    Parameters
+    ----------
+    max_iter : int, optional
+        The maximum number of iterations or equivalently the
+        the maximum number of solved subproblems.
+
+    max_epochs : int, optional
+        Maximum number of primal CD epochs on each subproblem.
+
+    dual_init : array, shape (n_samples,) default None
+        The initialization of dual variables.
+        If None, they are initialized as the 0 vector ``np.zeros(n_samples)``.
+
+    p0 : int, optional
+        First working set size.
+
+    tol : float, optional
+        The tolerance for the optimization.
+
+    verbose : bool or int, default False
+        Amount of verbosity. 0/False is silent.
+
+    References
+    ----------
+    .. [1] Olivier Fercoq and Pascal Bianchi,
+        "A Coordinate-Descent Primal-Dual Algorithm with Large Step Size and Possibly
+        Nonseparable Functions", SIAM Journal on Optimization, 2020,
+        https://epubs.siam.org/doi/10.1137/18M1168480,
+        code: https://github.com/Badr-MOUFAD/Fercoq-Bianchi-solver
+
+    .. [2] Bertrand, Q. and Klopfenstein, Q. and Bannier, P.-A. and Gidel, G.
+           and Massias, M.
+           "Beyond L1: Faster and Better Sparse Models with skglm", 2022
+           https://arxiv.org/abs/2204.07826
+    """
+
+    def __init__(self, max_iter=1000, max_epochs=1000, dual_init=None,
+                 p0=100, tol=1e-6, verbose=False):
+        self.max_iter = max_iter
+        self.max_epochs = max_epochs
+        self.dual_init = dual_init
+        self.p0 = p0
+        self.tol = tol
+        self.verbose = verbose
+
+    def solve(self, X, y, datafit_, penalty_, w_init=None, Xw_init=None):
+        if issparse(X):
+            raise ValueError("Sparse matrices are not yet support in PDCD_WS solver.")
+
+        datafit, penalty = PDCD_WS._validate_init(datafit_, penalty_)
+        n_samples, n_features = X.shape
+
+        # init steps
+        # Despite violating the conditions mentioned in [1]
+        # this choice of steps yield in practice a convergent algorithm
+        # with better speed of convergence
+        dual_step = 1 / norm(X, ord=2)
+        primal_steps = 1 / norm(X, axis=0, ord=2)
+
+        # primal vars
+        w = np.zeros(n_features) if w_init is None else w_init
+        Xw = np.zeros(n_samples) if Xw_init is None else Xw_init
+
+        # dual vars
+        if self.dual_init is None:
+            z = np.zeros(n_samples)
+            z_bar = np.zeros(n_samples)
+        else:
+            z = self.dual_init.copy()
+            z_bar = self.dual_init.copy()
+
+        p_objs = []
+        stop_crit = 0.
+        all_features = np.arange(n_features)
+
+        for iteration in range(self.max_iter):
+
+            # check convergence
```

Comment:
detail here: optimality criterion = max violation of primal and dual optimality conditions,
w in partial ... and z in partial ...

Reply:
in fact I realised it's the violation of the fixed point equation of Arrow Hurwicz for the primal, and the norm of the dual update for the dual. They make sense but they could be different, so it's worth mentioning in a comment here. this way the user doesn't have to go look at the functions